epimodel/tasks.py

```diff
@@ -0,0 +1,502 @@
+        regions = self.input()["region_file"].path
+        gleams = self.input()["gleam_regions"].path
+        aggregates = self.input()["aggregates"].path
+        rds = RegionDataset.load(regions, gleams, aggregates)
+        algorithms.estimate_missing_populations(rds)
+        with open(self.regions_dataset, "wb") as ofile:
+            dill.dump(rds, ofile)
+
+    def output(self):
+        return luigi.LocalTarget(self.regions_dataset)
+
+    def requires(self):
+        return {
+            "region_file": self.clone(RegionsFile),
+            "gleam_regions": self.clone(GleamRegions),
+            "aggregates": self.clone(RegionsAggregates),
+        }
+
+    @staticmethod
+    def load_dilled_rds(path: str):
+        with open(path, "rb") as ifile:
+            return dill.load(ifile)
+
+
+class JohnsHopkins(luigi.Task):
+    _output_directory: str = luigi.Parameter(default=CONFIG.output_directory)
+    hopkins_output: str = luigi.Parameter(
+        config_path=default_from_config("JohnsHopkins", "hopkins_output")
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._full_output_path = Path(self._output_directory) / self.hopkins_output
+
+    def run(self):
+        rds = RegionsDatasetTask.load_dilled_rds(self.input().path)
+        csse = imports.import_johns_hopkins(rds)
+        csse.to_csv(self._full_output_path)
+        logger.info(
+            f"Saved CSSE to {self._full_output_path}, last day is {csse.index.get_level_values(1).max()}"
+        )
+
+    def output(self):
+        return luigi.LocalTarget(self._full_output_path)
+
+    def requires(self):
+        return RegionsDatasetTask()
+
+
+class UpdateForetold(luigi.Task):
+    _output_directory: str = luigi.Parameter(default=CONFIG.output_directory)
+    foretold_output: str = luigi.Parameter(
+        config_path=default_from_config("UpdateForetold", "foretold_output")
+    )
+    foretold_channel: str = luigi.Parameter(
+        default="", description="The secret to fetch data from Foretold via API",
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._full_output_path = Path(self._output_directory) / self.foretold_output
+
+    def run(self):
+        if (
+            not isinstance(self.foretold_channel, str)
+            or len(self.foretold_channel) < 20
+        ):
+            raise ValueError(
+                "Foretold channel is either not a string or is too short to be valid"
+            )
+
+        logger.info("Downloading and parsing foretold")
+        rds = RegionsDatasetTask.load_dilled_rds(self.input().path)
+        foretold = imports.import_foretold(rds, self.foretold_channel)
+        foretold.to_csv(self._full_output_path, float_format="%.7g")
+        logger.info(f"Saved Foretold to {self._full_output_path}")
+
+    def output(self):
+        return luigi.LocalTarget(self._full_output_path)
+
+    def requires(self):
```

Comment:
Can we do it so that the order of the methods is always `requires`, `run`, `output` for scannability

Reply:
We can make it consistent at least. I would have slight preference for `requires`, `output`, `run` (even though it's not the "interpreter flow")

Reply:
:+1: I'm also fine your option